Fix test_mixed_par_ser hang caused by PR #906 access mode handling

brtnfld opened the pull request:
PR #906 (Fix #836) removed the hdf5_access_mode reset in cgp_close() to prevent corruption of parallel file operations when multiple files are open. However, this caused hdf5_access_mode to persist in PARALLEL state after cgp_open() calls, leading to deadlocks when subsequent cg_open() calls in MPI programs attempted collective HDF5 operations with only a subset of ranks participating.

The fix introduces an internal coordination flag (cgp_open_active) to distinguish between:
- Direct cg_open() calls: force NATIVE mode to avoid collective operations
- cg_open() called FROM cgp_open(): preserve PARALLEL mode for the file

Added additional testing, and added a C test.
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Fixes hang in MPI programs by managing `hdf5_access_mode` with a new `cgp_open_active` flag and adds tests for mixed parallel and serial I/O operations.
> 
>   - **Behavior**:
>     - Introduces `cgp_open_active` flag in `cgns_io.c` to manage `hdf5_access_mode` during `cg_open()` and `cgp_open()` calls.
>     - Ensures `cg_open()` forces NATIVE mode unless called from `cgp_open()`.
>     - Preserves PARALLEL mode for `cgp_open()` calls.
>   - **Testing**:
>     - Adds `test_mixed_par_ser.c` to test mixing of parallel and serial CGNS I/O operations.
>     - Updates `CMakeLists.txt` and `Makefile.in` to include new test.
>   - **Misc**:
>     - Renames `test_mixed_par_ser.F90` to `test_mixed_par_ser_f90.F90` for consistency.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=CGNS%2FCGNS&utm_source=github&utm_medium=referral)<sup> for ad0a6909d557baa3ee4a6b4be1be3299ae5f6d66. You can [customize](https://app.ellipsis.dev/CGNS/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

brtnfld committed:
Fix test_mixed_par_ser hang caused by PR #906 access mode handling

PR #906 (Fix #836) removed the hdf5_access_mode reset in cgp_close() to
prevent corruption of parallel file operations when multiple files are
open. However, this caused hdf5_access_mode to persist in PARALLEL state
after cgp_open() calls, leading to deadlocks when subsequent cg_open()
calls in MPI programs attempted collective HDF5 operations with only a
subset of ranks participating.

The fix introduces an internal coordination flag (cgp_open_active) to
distinguish between:
- Direct cg_open() calls: force NATIVE mode to avoid collective operations
- cg_open() called FROM cgp_open(): preserve PARALLEL mode for the file

Added additional testing, and added a C test.

brtnfld committed:
added missing test

brtnfld committed:
removed unused variable

brtnfld commented on src/cgnslib.c:
```diff
@@ -497,6 +500,9 @@
     double dummy_id;
     float FileVersion;
+#if CG_BUILD_HDF5
+    access_mode_t saved_access_mode;
```
fixed

brtnfld committed:
misc. cleanup

brtnfld committed:
cleanup

brtnfld committed:
fix compiler error

brtnfld commented:
@MicK7, can you review this when you get a chance?

MicK7 commented:
I understand the purpose of these changes. I am just wondering if the same result could be achieved without the cgp_open_active flag.
Indeed would it be equivalent to put the current code of cg_open in a private cgi_open that will have an API like that:
``` ier = cgi_open(char *filename, int mode, int open_parallel,  int *fn); ```
Then cgp_open will call cgi_open with the open_parallel to 1 while cg_open will call cgi_open with open_parallel at 0.

brtnfld committed:
Introduce internal cgi_open() function that accepts an explicit
open_parallel parameter to control HDF5 access mode, replacing
implicit global state coordination between cg_open() and cgp_open().

Implementation:
- Add cgi_open() internal function with open_parallel parameter
  - Parameter controls whether to preserve PARALLEL or force NATIVE mode
  - Contains core file opening logic for both serial and parallel paths

brtnfld commented:
> I understand the purpose of these changes. I am just wondering if the same result could be achieved without the cgp_open_active flag. Indeed would it be equivalent to put the current code of cg_open in a private cgi_open that will have an API like that: `ier = cgi_open(char *filename, int mode, int open_parallel, int *fn);` Then cgp_open will call cgi_open with the open_parallel to 1 while cg_open will call cgi_open with open_parallel at 0.

Thanks for the suggestion, I changed it to do this.
